src/server/pkg/storage/chunk/reader.go

```diff
@@ -116,90 +190,56 @@ func (r *Reader) readChunk(chunk *Chunk) error {
 		return err
 	}
 	defer gzipR.Close()
-	r.buf.Reset()
-	buf := grpcutil.GetBuffer()
-	defer grpcutil.PutBuffer(buf)
-	if _, err := io.CopyBuffer(r.buf, gzipR, buf); err != nil {
+	buf := &bytes.Buffer{}
+	if _, err := io.Copy(buf, gzipR); err != nil {
 		return err
 	}
+	dr.chunk = buf.Bytes()
 	return nil
 }
 
-// OnSplit registers a callback for when a chunk split point is encountered.
-// The callback is only executed at a split point found after reading WindowSize bytes.
-// The reason for this is to guarantee that the same split point will appear in the writer
-// the data is being written to.
-func (r *Reader) OnSplit(f func()) {
-	r.bytesBeforeSplit = 0
-	r.onSplit = f
+// BeforeBound checks if the passed in string is before the string bound (exclusive).
+// The string bound is optional, so if no string bound is passed then it returns true.
+func BeforeBound(str string, strBound ...string) bool {
+	return len(strBound) == 0 || strings.Compare(str, strBound[0]) < 0
 }
 
-func (r *Reader) executeOnSplitFunc() {
-	if r.onSplit != nil && r.bytesBeforeSplit > WindowSize {
-		r.onSplit()
-		r.onSplit = nil
+// Get writes the data referenced by the data reference.
+func (dr *DataReader) Get(w io.Writer) error {
+	if err := dr.getChunk(); err != nil {
+		return err
 	}
+	data := dr.chunk[dr.dataRef.OffsetBytes : dr.dataRef.OffsetBytes+dr.dataRef.SizeBytes]
+	if _, err := w.Write(data); err != nil {
+		return err
+	}
+	return nil
 }
 
-// Copy is the basic data structure to represent a copy of data from
-// a reader to a writer.
-// before/after are the raw bytes that precede/follow full chunks
-// in the set of bytes represented by the copy.
-type Copy struct {
-	before, after *bytes.Buffer
-	chunkRefs     []*DataRef
-}
-
-// ReadCopy reads copy data from the reader.
-func (r *Reader) ReadCopy(n ...int64) (*Copy, error) {
-	totalLeft := int64(math.MaxInt64)
-	if len(n) > 0 {
-		totalLeft = n[0]
-		if r.Len() < totalLeft {
-			return nil, fmt.Errorf("reader length (%v) less than copy length (%v)", r.Len(), totalLeft)
-		}
-	}
-	rawCopy := func(w io.Writer, n int64) error {
-		if _, err := io.CopyN(w, r, n); err != nil {
-			return err
+// LimitReader progresses the data reader to a tag bound and will return
+// a new data reader that will read the tags up to the tag bound. This does
```

Comment:
Little ambiguous here which tag bound is the `tagBound` argument below (I think it's the second), and which is the previous tag bound. Would be good to clarify this a bit.

Reply:
There is only one tag bound involved here. A data reader has a set of tags it can iterate through, and creating a limit reader will create a new data reader that can read up to the tag bound from that set (while also progressing the original data reader past those tags). I am going to re-word this a bit and see if that helps.

Reply:
Makes sense, I think it could be a bit clearer, but looking at it now I'm not sure where I was getting 2 tagbounds from.